test/e2e/instrumentation/monitoring/prometheus.go

```diff
@@ -0,0 +1,393 @@
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/prometheus/common/model"
+
+	. "github.com/onsi/ginkgo"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/common"
+	"k8s.io/kubernetes/test/e2e/framework"
+	instrumentation "k8s.io/kubernetes/test/e2e/instrumentation/common"
+)
+
+const (
+	prometheusQueryStep                = time.Minute
+	prometheusMetricErrorTolerance     = 0.25
+	prometheusMetricValidationDuration = time.Minute * 2
+	prometheusRate                     = time.Minute * 2
+	prometheusRequiredNodesUpDuration  = time.Minute * 5
+	prometheusService                  = "prometheus"
+	prometheusSleepBetweenAttempts     = time.Second * 30
+	prometheusTestTimeout              = time.Minute * 5
+	resourceConsummerDefaultCPU        = 10
+)
+
+var _ = instrumentation.SIGDescribe("[Feature:PrometheusMonitoring] Prometheus", func() {
+	BeforeEach(func() {
+		framework.SkipUnlessClusterMonitoringModeIs("prometheus")
+	})
+
+	f := framework.NewDefaultFramework("prometheus-monitoring")
+	It("should scrape container metrics from all nodes.", func() {
+		expectedNodes, err := getAllNodes(f.ClientSet)
+		framework.ExpectNoError(err)
+		retryUntilSucceeds(func() error {
+			return validateMetricAvailableForAllNodes(f.ClientSet, `container_cpu_usage_seconds_total`, expectedNodes)
+		}, prometheusTestTimeout)
+	})
+	It("should successfully scrape all targets", func() {
+		retryUntilSucceeds(func() error {
+			return validateAllActiveTargetsAreHealthy(f.ClientSet)
+		}, prometheusTestTimeout)
+	})
+	It("should contain correct container CPU metric.", func() {
+		var targetCPUUsage float64 = 0.1
+		var requiredSampleCount int = 3
+		var consumerName string = "prometheus-cpu-consumer"
+		query := prometheusCPUQuery(f.Namespace.Name, consumerName, prometheusRate)
+		consumer := consumeCPUResources(f, consumerName, int(targetCPUUsage*1000))
+		defer consumer.CleanUp()
+		retryUntilSucceeds(func() error {
+			return validateQueryReturnsCorrectValues(f.ClientSet, query, targetCPUUsage, requiredSampleCount, prometheusMetricErrorTolerance)
+		}, prometheusTestTimeout)
+	})
+	It("should scrape metrics from annotated pods.", func() {
+		var customMetricTargetValue float64 = 1000
+		var consumerName string = "prometheus-custom-pod-metric"
+		var requiredSampleCount int = 1
+		var annotatePods, annotateService bool = true, false
+		query := prometheusPodCustomMetricQuery(f.Namespace.Name, consumerName)
+		consumer := exportCustomMetric(f, consumerName, int(customMetricTargetValue), annotatePods, annotateService)
+		defer consumer.CleanUp()
+		retryUntilSucceeds(func() error {
+			return validateQueryReturnsCorrectValues(f.ClientSet, query, customMetricTargetValue, requiredSampleCount, prometheusMetricErrorTolerance)
+		}, prometheusTestTimeout)
+	})
+	It("should scrape metrics from annotated services.", func() {
+		var customMetricTargetValue float64 = 1000
```

Comment:
I think you missed my previous comment - can you use lliterals instead? (you can use bool constants for readability: doAnnotatePods = true, doAnnotateServices = true, dontAnnotatePods = false, dontAnnotateServices = false)

Reply:
I moved test parameterization to consts and split `exposeCustomMetrics` function based on annotation type.